test/hotspot/jtreg/compiler/jvmci/jdk.vm.ci.hotspot.test/src/jdk/vm/ci/hotspot/test/TestDynamicConstant.java

```diff
@@ -0,0 +1,304 @@
+/*
+ * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @requires vm.jvmci
+ * @summary Test CONSTANT_Dynamic resolution by HotSpotConstantPool.
+ * @modules java.base/jdk.internal.org.objectweb.asm
```

Comment:
Instead of ASM, I would preferred to use `test.java.lang.invoke.lib.InstructionHelper` like `CondyBSMInvocation` [does](https://github.com/openjdk/jdk/blob/05a764f4ffb8030d6b768f2d362c388e5aabd92d/test/jdk/java/lang/invoke/condy/CondyBSMInvocation.java#L56). However, I could not figure out how to use a test library from the `test/jdk` suite in the `test/hotspot/jtreg` suite. Do you know if that's possible @PaulSandoz ?

Reply:
It might be (depending of source locations). I would need to look in more detail. I can do that as a follow on investigation/fix if need be, rather than block this PR?

Reply:
No need to spend time on this - just thought you may know off the top of your head. I'll investigate further myself.
Thanks for the detailed review!